Xamarin.Forms.Core/Shell/ShellUriHandler.cs

```diff
@@ -274,124 +263,207 @@ internal static List<RouteRequestBuilder> GenerateRoutePaths(Shell shell, Uri re
 
 			if (relativeMatch && shell?.CurrentItem != null)
 			{
-				// retrieve current location
-				var currentLocation = NodeLocation.Create(shell);
+				var result = ProcessRelativeRoute(shell, routeKeys, segments, enableRelativeShellRoutes, originalRequest);
+				if (result.Count > 0)
+					return result;
+			}
 
-				while (currentLocation.Shell != null)
-				{
-					var pureRoutesMatch = new List<RouteRequestBuilder>();
-					var pureGlobalRoutesMatch = new List<RouteRequestBuilder>();
+			possibleRoutePaths.Clear();
+			SearchPath(shell, null, segments, possibleRoutePaths, depthStart);
 
-					//currently relative routes to shell routes isn't supported as we aren't creating navigation stacks
-					if (enableRelativeShellRoutes)
-					{
-						SearchPath(currentLocation.LowestChild, null, segments, pureRoutesMatch, 0);
-						ExpandOutGlobalRoutes(pureRoutesMatch, routeKeys);
-						pureRoutesMatch = GetBestMatches(pureRoutesMatch);
-						if (pureRoutesMatch.Count > 0)
-						{
-							return pureRoutesMatch;
-						}
-					}
+			var bestMatches = GetBestMatches(possibleRoutePaths);
+			if (bestMatches.Count > 0)
+				return bestMatches;
 
+			bestMatches.Clear();
+			ExpandOutGlobalRoutes(possibleRoutePaths, routeKeys);
 
-					SearchPath(currentLocation.LowestChild, null, segments, pureGlobalRoutesMatch, 0, ignoreGlobalRoutes: false);
-					ExpandOutGlobalRoutes(pureGlobalRoutesMatch, routeKeys);
+			foreach(var possibleRoutePath in possibleRoutePaths)
+			{
+				if (possibleRoutePath.IsFullMatch)
+					continue;
+				
+				var url = possibleRoutePath.PathFull;
+				
+				var globalRouteMatches = 
+					SearchForGlobalRoutes(
```

Comment:
This is new code

Same description as above.  Once we get here if the full route hasn't been matched then we try to fulfill the rest of the request with global registered routes.